Switch autodeployment server to use streaming cache for build logs

humphd opened the pull request:
Fixes #2695
I noticed a bug with the way I was caching the build logs: the cached contents get erased the first time they are requested.  So it was working the first time, but not on the second, third, ... request.

This switches the log cache to use an LRU stream cache, see https://github.com/LaurentZuijdwijk/streaming-cache.

It also lets us hold on to more build logs from the past based on git sha.

AmasiaNalbandian reviewed (dismissed):
the code looks good! I'm just not sure how we can confirm it's working.... Also can you please resolve conflicts for the pnpm package?

humphd commented:
@manekenpix has to manually deploy this (it doesn't build automatically), so when he has time to try it, and can go for it.  I'll update the lock file.

AmasiaNalbandian commented on pnpm-lock.yaml:
```diff
@@ -401,7 +401,7 @@ importers:
       pm2: ^5.1.1
       shelljs: ^0.8.4
-      stream-buffers: ^3.0.2
+      streaming-cache: ^0.5.5
```
Do we want the carat for streaming-cache? or should we manually keep track of it?

humphd replied:
We need to pin all of these, which is what #2725 is about.

manekenpix commented on tools/autodeployment/builds.js:
```diff
@@ -1,13 +1,26 @@
     this.startedDate = new Date();
+
+    // Remember the git SHA for this build
+    this.sha = githubData.after;
```
I think this won't work in production. I left a [comment in 2650](https://github.com/Seneca-CDOT/telescope/issues/2650#issuecomment-1018045662) for a possible fix.

manekenpix commented:
I'm not sure if this is work fixing here, but [this](https://github.com/Seneca-CDOT/telescope/blob/ef163e4e87e93eeb8cba4ffc11e8042b99c5fe2a/tools/autodeployment/builds.js#L40) is preventing the `autodeployment` server from sending the status code when the build succeeds. When the status code is 0, it's not copied into the `build` object, and `code` is missing in the data fetched by the dashboard.

humphd commented:
I've reworked this, and it's ready for another review.  @manekenpix, thanks for testing it on staging.  I've fixed the crashes you found.  I also needed to switch the streaming cache logic to use a re-readable stream, since the way I was doing it was still consuming the stream after the first read.

I've tested this with a hacked up local autodeployment server, and I think it's working. now.  If you could test it again on staging, I'd appreciate it.

humphd commented:
Prettier failures are due to issues on `master` that will get fixed and I'll rebase.

DukeManh commented:
@humphd, can you rebase, please.

humphd commented:
Squashed and Rebased.

manekenpix commented:
@humphd I'm ok if you want to rebase and merge.

humphd commented:
Hurray 🎆
